lib/pbench/server/cache_manager.py

```diff
@@ -490,52 +410,54 @@ def __init__(self, path: Path, incoming: Path, results: Path, logger: Logger):
         # The full controller path in the ARCHIVE tree
         self.path = path
 
+        # Remember the cache tree base
+        self.cache = cache
+
         # Provide a mapping from Tarball file name to object
-        self.tarballs: Dict[str, Tarball] = {}
+        self.tarballs: dict[str, Tarball] = {}
 
         # Provide a mapping from Dataset resource ID to object
-        self.datasets: Dict[str, Tarball] = {}
-
-        # The directory where the controller's tarballs will be unpacked
-        self.incoming: Path = incoming / self.name
+        self.datasets: dict[str, Tarball] = {}
 
-        # A path that will link to the controller's unpacked tarballs
-        self.results: Path = results / self.name
+        # Discover the tarballs that already exist
+        # DEPENDS ON self.cache!
         self._discover_tarballs()
 
     def _discover_tarballs(self):
-        """
-        Discover the tarballs within the ARCHIVE tree's controller directory.
-        Check for an unpacked tarball in the INCOMING tree and if that's
-        present also check for a RESULTS tree link.
+        """Discover the known tarballs
+
+        Look in the ARCHIVE tree's controller directory for tarballs, and add
+        them to the known set. We also check for unpacked directories in the
+        CACHE tree matching the resource_id of any tarballs we find in order
+        to link them.
+
+        Args:
+            cache:  The cache tree base for temporarily unpacked tarballs
```

Comment:
I don't see where `cache` is an argument to this function (it's not part of the function signature, and `self.cache` (despite the assertion at line 423) isn't directly referenced in the body).

That said, `Controller.cache` **is** referenced inside the `Tarball` c'tor, but that's pretty "far away" to be documented here.  (And, it's not the only `Controller` attribute referenced there....)

So, I recommend removing lines 433-435.

I also recommend recasting line 423 to the standard disclaimer that we use when we make an instance method call inside a c'tor -- that the subroutine depends on attributes of the instance, and therefore all initialization should be complete before making the call.

Reply:
Sigh; continuous refactoring somehow left the docstring behind. It was probably hiding with the other one.